pkg/kv/kvserver/apply/asserter.go

```diff
@@ -0,0 +1,411 @@
+	// original seed proposal.
+	if seedID != "" {
+		if seedProposedBy, ok := r.proposedCmds[seedID]; !ok {
+			// INVARIANT: a LAI reproposal must reference a previous seed proposal.
+			fail("unknown seed proposal %s", seedID)
+		} else if seedProposedBy != replicaID {
+			// INVARIANT: a LAI reproposal must be made by the seed replica.
+			fail("seed proposal %s by different replica %d", seedID, seedProposedBy)
+		}
+
+		if s, ok := r.seedCmds[cmdID]; ok && s != seedID {
+			// INVARIANT: a reproposal of a LAI reproposal must always reference the
+			// same seed proposal.
+			fail("expected seed proposal %s, got %s", s, seedID)
+		}
+		r.seedCmds[cmdID] = seedID
+	}
+}
+
+// Apply tracks and asserts command application. Rejected commands (e.g. via LAI
+// checks) are not considered applied. The command must be reported via
+// Propose() first.
+func (a *Asserter) Apply(
+	rangeID roachpb.RangeID,
+	replicaID roachpb.ReplicaID,
+	cmdID kvserverbase.CmdIDKey,
+	entry raftpb.Entry,
+	leaseAppliedIndex kvpb.LeaseAppliedIndex,
+	closedTS hlc.Timestamp,
+) {
+	a.forRange(rangeID).apply(replicaID, cmdID, entry, leaseAppliedIndex, closedTS)
+}
+
+func (r *rangeAsserter) apply(
+	replicaID roachpb.ReplicaID,
+	cmdID kvserverbase.CmdIDKey,
+	raftEntry raftpb.Entry,
+	leaseAppliedIndex kvpb.LeaseAppliedIndex,
+	closedTS hlc.Timestamp,
+) {
+	entry := applyState{
+		cmdID:             cmdID,
+		raftIndex:         kvpb.RaftIndex(raftEntry.Index),
+		raftTerm:          kvpb.RaftTerm(raftEntry.Term),
+		leaseAppliedIndex: leaseAppliedIndex,
+		closedTS:          closedTS,
+	}
+
+	fail := func(msg string, args ...interface{}) {
+		panic(fmt.Sprintf("r%d/%d: %s (%s)\ndata: %x",
+			r.rangeID, replicaID, fmt.Sprintf(msg, args...), entry, raftEntry.Data))
+	}
+
+	// INVARIANT: all commands have a command ID. etcd/raft may commit noop
+	// proposals on leader changes that do not have a command ID, but we skip
+	// these during application.
+	if len(cmdID) == 0 {
+		fail("applied command without ID")
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	// INVARIANT: a command must be proposed before it can apply.
+	if _, ok := r.proposedCmds[cmdID]; !ok {
+		fail("command was not proposed")
+	}
+
+	if int(entry.raftIndex) < len(r.log) {
+		// INVARIANT: all replicas must apply the same log entry at the same index.
+		if e := r.log[int(entry.raftIndex)]; e != entry {
+			fail("applied entry differs from existing log entry: %s", e)
+		}
+
+	} else {
+		// The entry is not yet tracked in the log.
+
+		// INVARIANT: entries may not regress the applied state.
+		if len(r.log) > 0 {
+			s := r.stateAt(kvpb.RaftIndex(len(r.log) - 1))
+			// INVARIANT: the Raft index must progress.
+			if entry.raftIndex <= s.raftIndex {
+				fail("Raft index regression %d -> %d", s.raftIndex, entry.raftIndex)
+			}
+			// INVARIANT: the Raft term must not regress.
+			if entry.raftTerm < s.raftTerm {
+				fail("Raft term regression %d -> %d", s.raftTerm, entry.raftTerm)
+			}
```

Comment:
Yep, but that's the case for many other assertions here as well (e.g. that all replicas apply the same command at the same index). We may as well assert these invariants here as well, since we rely on them holding. Don't feel particularly strongly about it though, but it doesn't really cost us anything either.